internal/metrics/daemon.go

```diff
@@ -0,0 +1,202 @@
+package metrics
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"git.sr.ht/~spc/go-log"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	workloadMetric = "workloadName"
+	scrapeTimeout  = 15 * time.Second
+)
+
+type ScrappeCallback func(data []model.Vector, labels map[string]string)
+
+type TargetMetric struct {
+	workloadName string
+	interval     time.Duration
+	urls         []string
+	scrappers    []*HTTPScraper
+	ctx          context.Context
+	cancel       context.CancelFunc
+	forcetrigger chan bool
+	lock         sync.RWMutex
+}
+
+func NewTargetMetric(workloadName string, interval time.Duration, urls []string) *TargetMetric {
+	return &TargetMetric{
+		workloadName: workloadName,
+		interval:     interval,
+		urls:         urls,
+		forcetrigger: make(chan bool),
+	}
+}
+func (tg *TargetMetric) ForceEvent() {
+	tg.forcetrigger <- true
+}
+
+func (tg *TargetMetric) Start(cb ScrappeCallback) {
+	log.Infof("started targetMetric for workload '%v' with urls '%+v'",
+		tg.workloadName, tg.urls)
+
+	tg.lock.Lock()
+	// if it was started early, stop it
+	if tg.cancel != nil {
+		tg.cancel()
+	}
+
+	tg.createScrapper()
+	tg.ctx, tg.cancel = context.WithCancel(context.Background())
+	ticker := time.NewTicker(tg.interval)
+	tg.lock.Unlock()
+
+	getdata := func() {
+		log.Debugf("ticker run for workload '%v'", tg.workloadName)
+		ctx, _ := context.WithTimeout(context.Background(), scrapeTimeout)
+		data := tg.Run(ctx)
+		cb(data, map[string]string{workloadMetric: tg.workloadName})
```

Comment:
WDYT about changning it into something more generic, like `metricSource` that could be both workload name or "system"?
We will also need to add device id to each of the metrics, but that I think can be done when the metrics are sent to the control plane.